plugin/src/main/java/io/fabric8/maven/plugin/BuildMojo.java

```diff
@@ -241,7 +293,191 @@ private void executeOpenShiftBuild(ServiceHub hub, ImageConfiguration imageConfi
+
+        waitForOpenShiftBuildToComplete(client, buildName, build);
+
+        updateImageStreamTags(client, imageConfig, buildName, build);
+    }
+
+
+    private void waitForOpenShiftBuildToComplete(OpenShiftClient client, String buildConfigName, Build build) {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        String buildName = getName(build);
+        Watcher<Build> buildWatcher = new Watcher<Build>() {
+            @Override
+            public void eventReceived(Action action, Build resource) {
+                if (isBuildCompleted(action, resource)) {
+                    latch.countDown();
+                }
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
+            }
+        };
+        log.info("Waiting for build " + buildName + " to complete...");
+        try (Watch watcher = client.builds().withName(buildName).watch(buildWatcher)) {
+            while (latch.getCount() > 0L) {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    // ignore
+                }
+            }
+            log.info("Build " + buildName + " completed!");
+        }
+    }
+
+
+    /**
+     * Lets update the ImageStream to include the correct tags
+     */
+    private void updateImageStreamTags(OpenShiftClient client, ImageConfiguration imageConfig, String buildConfigName, Build build) throws MojoExecutionException {
+
+        try {
+            File manifest = openshiftManifest;
+            if (!Files.isFile(manifest)) {
+                throw new MojoFailureException("No such generated manifest file: " + manifest);
+            }
+
+            String namespace = clusterAccess.getNamespace();
+            Controller controller = new Controller(client);
+
+            boolean updated = false;
+            Set<HasMetadata> entities = loadResources(client, controller, namespace, manifest, project, log);
+            for (HasMetadata entity : entities) {
+                if (entity instanceof ImageStream) {
+                    ImageStream is = (ImageStream) entity;
+                    String imageStreamName = KubernetesHelper.getName(is);
+                    if (Objects.equals(buildConfigName, imageStreamName)) {
+                        if (updateImageStreamTag(client, imageConfig, is, buildConfigName)) {
+                            updated = true;
+                        }
+                    }
+                }
+            }
+            if (updated) {
+                // lets store the entities again!
+                KubernetesList entity = new KubernetesListBuilder().withItems(new ArrayList<>(entities)).build();
+                File resourceFileBase = new File(this.targetDir, ResourceClassifier.OPENSHIFT.getValue());
+                AbstractResourceMojo.writeResourcesIndividualAndComposite(entity, resourceFileBase, ResourceFileType.yaml, log);
```

Comment:
> To simplify things I'm tempted to refactor the logic along these lines; that ResourceMojo generates most resources; then BuildMojo appends ImageStreams in openshift mode; then they are logically combined - I'd also like to make them order agnostic (so re-running ResourceMojo in isolation won't zap all the ImageStream resources generated from the previous BuildMojo goal).

That's a very good idea, so that we can run both goals independently if they write into different files. For the uber jar (which we don't need for running `fabric8:deploy`) we could use an own `fabric8:package`. Then the analogy to the Maven lifecycyle would be complete:

* *resource* --> `fabric8:resource`
* *compile* --> `fabric8:build`
* *package* --> `fabric8:package`

`fabric8:package` would then be responsible for packing up things like creating the uber yaml. That way we can let `fabric8:resource` and `fabric8:build` write in to separate files, `fabric8:package` would then combine them (and if required, take care of the order).

BTW, there is one thing were ImageStream are superior than images in registries: With refering to image in registries we also bind the resource objects (pod spec) to a specific registry, which would be the same as say, you need to refer to a maven repo from within a war.xml in an war. This means you refer to meta information which might change (e.g. when you move your app image to another registry you need to update the resource descriptor within the app). 

With `ImageStreams` (not *ImageStreamTag* !) you abstract away the registry (its the internal registry which can delegate to other registries, too), so in principle this is better for an universal resource descriptor. Not for ImageStreamTag refering to a very specific image via a checksum. Maybe we can even work more with ImageStreams themself ? (not so sure what I'm saying here ;-)